lib/internal/modules/esm/import_map.js

```diff
@@ -0,0 +1,173 @@
+    let subSpec;
+    let bareSpec;
+    if (isURL(spec)) {
+      spec = spec.href;
+    } else if (hasSlash !== -1) {
+      hasSlash += 1;
+      subSpec = StringPrototypeSlice(spec, hasSlash);
+      bareSpec = StringPrototypeSlice(spec, 0, hasSlash);
+    }
+
+    let _mappedSpecifier = this.#imports.get(bareSpec) || this.#imports.get(spec);
+    if (_mappedSpecifier) {
+      // Re-assemble sub spec
+      if (_mappedSpecifier === spec && subSpec) {
+        _mappedSpecifier += subSpec;
+      }
+      const mappedSpecifier = this.#getMappedSpecifier(_mappedSpecifier);
+
+      if (mappedSpecifier !== _mappedSpecifier) {
+        this.imports.set(specifier, mappedSpecifier);
+      }
+      specifier = mappedSpecifier;
+    }
+
+    return specifier;
+  }
+
+  process(raw) {
+    if (!raw) {
+      throw new ERR_INVALID_IMPORT_MAP('top level must be a plain object');
+    }
+
+    // Validation and normalization
+    if (raw.imports === null || typeof raw.imports !== 'object' || ArrayIsArray(raw.imports)) {
+      throw new ERR_INVALID_IMPORT_MAP('top level key "imports" is required and must be a plain object');
+    }
+    if (raw.scopes === null || typeof raw.scopes !== 'object' || ArrayIsArray(raw.scopes)) {
+      throw new ERR_INVALID_IMPORT_MAP('top level key "scopes" is required and must be a plain object');
+    }
+
+    // Normalize imports
+    const importsEntries = ObjectEntries(raw.imports);
+    for (let i = 0; i < importsEntries.length; i++) {
+      const { 0: specifier, 1: mapping } = importsEntries[i];
+      if (!specifier || typeof specifier !== 'string') {
+        throw new ERR_INVALID_IMPORT_MAP('module specifier keys must be non-empty strings');
+      }
+      if (!mapping || typeof mapping !== 'string') {
+        throw new ERR_INVALID_IMPORT_MAP('module specifier values must be non-empty strings');
+      }
+      if (StringPrototypeEndsWith(specifier, '/') && !StringPrototypeEndsWith(mapping, '/')) {
+        throw new ERR_INVALID_IMPORT_MAP('module specifier keys ending with "/" must have values that end with "/"');
+      }
+
+      this.imports.set(specifier, mapping);
+    }
+
+    // Normalize scopes
+    // Sort the keys according to spec and add to the map in order
+    // which preserves the sorted map requirement
+    const sortedScopes = ArrayPrototypeReverse(ArrayPrototypeSort(ObjectKeys(raw.scopes)));
+    for (let i = 0; i < sortedScopes.length; i++) {
+      let scope = sortedScopes[i];
+      const _scopeMap = raw.scopes[scope];
+      if (!scope || typeof scope !== 'string') {
+        throw new ERR_INVALID_IMPORT_MAP('import map scopes keys must be non-empty strings');
+      }
+      if (!_scopeMap || typeof _scopeMap !== 'object') {
+        throw new ERR_INVALID_IMPORT_MAP(`scope values must be plain objects (${scope} is ${typeof _scopeMap})`);
+      }
+
+      // Normalize scope
+      debugger
+      scope = new URL(scope, this.#baseURL);
```

Comment:
NIT FYI: Performance-wise this whole class should be implemented in C++.

Reply:
I think we should wait until we are stable before doing this. Much easier to iterate in JS. There are also a few things to be done to improve the resolve perf (using a radix tree for the scope matching specifically) that we can do before deciding if this is hot enough to need c++ imo, but I still think we should wait to do that until later.

Reply:
Import maps do linear largest prefix not radix random lookup on trailing / so you have to linear search it. That's more complexity than I'd desire in a single PR.

Reply:
I agree and is why I went with the initial approach. I just meant to say that if we are doing a pref review, the intent of this PR is not to do it the most performant way.

Reply:
Calling this resolved for now. I agree though, we should take a performance pass once we have stabilized this new api.